### PR TITLE
Tabulator tests

### PR DESCRIPTION
**Pull request by mikefranze**
Adding basic tests for approval, plurality, ranked robin, and IRV. 

There's probably some edge cases I missed but this adds the framework to add more later.

### CODE REVIEW

File: backend/src/Tabulators/Approval.test.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
+import { Approval } from './Approval'
+
+describe("Approval Tests", () => {
+    test("Single Winner Test", () => {
+        // Simple test to elect the candidate with most votes
+        const candidates = ['Alice', 'Bob', 'Carol', 'Dave']
+
+        const votes = [
+            [1, 1, 1, 1],
```
Comment (ArendPeter) [inline backend/src/Tabulators/Approval.test.ts:9]: At the moment only ``0, 0, 0, 0`` is considered an under vote, but ``1, 1, 1, 1`` has the same effect on the election results, so we could also consider classifying that as an undervote

File: backend/src/Tabulators/Approval.test.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        expect(results.summaryData.totalScores[0].index).toBe(3)
+        expect(results.summaryData.totalScores[1].score).toBe(6)
+        expect(results.summaryData.totalScores[1].index).toBe(2)
+        expect(results.summaryData.totalScores[2].score).toBe(3)
+        expect(results.summaryData.totalScores[2].index).toBe(1)
+        expect(results.summaryData.totalScores[3].score).toBe(1)
+        expect(results.summaryData.totalScores[3].index).toBe(0)
+        
+        expect(results.summaryData.nUnderVotes).toBe(1)
+        expect(results.summaryData.nValidVotes).toBe(7)
```
Comment (ArendPeter) [inline backend/src/Tabulators/Approval.test.ts]: Right now this doesn't include the under votes, but maybe it should?

File: backend/src/Tabulators/Approval.test.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        ]
+        const results = Approval(candidates, votes)
+        expect(results.elected.length).toBe(1);
+        expect(results.elected[0].name).toBe('Dave');
+        expect(results.summaryData.totalScores[0].score).toBe(7)
+        expect(results.summaryData.totalScores[0].index).toBe(3)
+        expect(results.summaryData.totalScores[1].score).toBe(6)
+        expect([1,2]).toContain(results.summaryData.totalScores[1].index)
+        expect(results.summaryData.totalScores[2].score).toBe(6)
+        expect([1,2]).toContain(results.summaryData.totalScores[2].index)
```
Comment (ArendPeter) [inline backend/src/Tabulators/Approval.test.ts:89]: nit: Maybe add a comment here indicating that random tiebreaker is the expected behavior that we're testing for

File: backend/src/Tabulators/IRV.test.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        const votes = [
+            [1, 2, 3],
+            [1, 2, 3],
+            [1, 2, 3],
+            [1, 2, 3],
+            [3, 2, 1],
+            [3, 2, 1],
+            [3, 2, 1],
+            [3, 2, 1],
+            [2, 1, 1],//overvote
```
Comment (ArendPeter) [inline backend/src/Tabulators/IRV.test.ts:55]: Nice job including all the edge cases :)

File: backend/src/Tabulators/IRV.test.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            [2, 1, 2],//second round overvote
+            [0, 1, 0],//second round exhausted vote
+        ]
+        const results = IRV(candidates, votes)
+        expect(results.elected[0].name).toBe('Alice');
+        expect(results.voteCounts.length).toBe(2);  //Two rounds
+        expect(results.voteCounts[0]).toStrictEqual([4,3,4]);  
+        expect(results.voteCounts[1]).toStrictEqual([5,0,4]);  
+        expect(results.overVoteCounts).toStrictEqual([1,2]);   
+        expect(results.exhaustedVoteCounts).toStrictEqual([1,3]);
```
Comment (ArendPeter) [inline backend/src/Tabulators/IRV.test.ts:66]: I think a comment would be helpful here clarifying which votes are overvoted / exhausted in each round. Here's my understanding

### Round 1
``[2, 1, 1],//overvote`` is over voted & exhausted

### Round 2
``[2, 1, 2],//second round overvote`` is over voted & exhausted
``[0, 1, 0],//second round exhausted vote`` is exhausted, but not overvoted

File: backend/src/Tabulators/Plurality.test.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        expect(results.summaryData.totalScores[2].index).toBe(1)
+        expect(results.summaryData.totalScores[3].score).toBe(0)
+        expect(results.summaryData.totalScores[3].index).toBe(0)
+        
+        expect(results.summaryData.nUnderVotes).toBe(1)
+        expect(results.summaryData.nValidVotes).toBe(11)
+        expect(results.summaryData.nInvalidVotes).toBe(3)
+    })
+
+    test("Multi Winner Test", () => {
```
Comment (ArendPeter) [inline backend/src/Tabulators/Plurality.test.ts:41]: This looks like it's still single winner?

File: backend/src/Tabulators/Plurality.test.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            [0, 0, 0, 1],
+            [0, 0, 0, 1],
+        ]
+        const results = Plurality(candidates, votes)
+        expect(results.elected.length).toBe(1);
+        expect(results.elected[0].name).toBe('Dave');
+        expect(results.summaryData.totalScores[0].score).toBe(5)
+        expect(results.summaryData.totalScores[0].index).toBe(3)
+        expect(results.summaryData.totalScores[1].score).toBe(3)
+        expect([1,2]).toContain(results.summaryData.totalScores[1].index)
```
Comment (ArendPeter) [inline backend/src/Tabulators/Plurality.test.ts:97]: same as with previous tie scenario, it would be good to add a comment clarifying that we're expecting the tie to be broken randomly

**Review by ArendPeter — APPROVED** [review]
Looks good, thanks for adding these tests!

**Review by ArendPeter — APPROVED** [review]
Re approved!
